Add Secure Storage

tomjackman opened the pull request:
Add Secure Storage with Realm Mobile DB and CRUDL functionality.


Code changes added to this commit for easier review - add realm db storage (f257637)

Some notes:
- Edit/Delete options are shown when swiping slightly left on a note in the list (full swipe across will also delete the note quickly).
- Completion handlers were added for the CRUDL actions on the Realm db but were removed as they didn't provide much benefits since the the realm db will only give errors like: bad realm config, wrong decryption key, rather then return an error if the object you want to delete doesn't exist in the db, which is probably the cases we would really want.

To Do's:
- [x] Move Add note menu bar button to override/merge with the global nav bar

Commit:
add realm db storage

wei-lee commented:
It works pretty well when I test it on the simulator. Let's try it on a real device and see what's the performance like

tomjackman commented:
- Added calls to the realm db to happen on a background thread, UI updates (if any) are done in the ui thread.
- Removed TableViewCell class file and moved it to the StorageViewController instead. (Unnecessary functions removed - super.x() calls)
- Fixed bug in the note id generator - it was generating a number less than 64, rather than a 64 byte string for the ID.
- Fixed indentation
- Made necessary function private

I tested in the emulator creating large numbers of notes (100) back to back and they get created very quickly. Similarly with a large table of notes (around 250-300) they were rendered instantly. Realm seems to do some caching of objects which might help.

wei-lee commented on secure-ios-app/storage/StorageViewController.swift:
```diff
@@ -192,9 +197,15 @@ class StorageViewController: UITableViewController {
+                self.notes = storedNotes!
+            }
+            DispatchQueue(label: "ui").sync {
+                self.tableView.reloadData()
```
I see here another 2 instances of `DispatchQueue` are created and each of them is processing some task. In this case, I think it's not guaranteed that the notes will be the latest when the `tableView` is reloaded. Not sure if that is intended?

wei-lee replied:
Also I don't think the tasks submitted to the `ui` queue are not actually executed on the main UI thread, it's a bit confusing.

tomjackman commented:
Tried to improve the threading with realm. Realm isn't very thread friendly - any data you retrieve from the database on one thread, you can't pass to another thread. Also realm instances can't be used across threads either. This is a design choice in realm. In the lastest commit, I have omitted the read (read note, list notes) db calls from happening on the background thread (since we can't pass realm originating data on the background thread to the main UI thread). However, the remaining calls (create, edit, delete, delete all) occur on the background thread.

An overview of the general approach that was taken:
- On the main thread in the view controller, call the storage service to create a note.
- The storage service creates a new thread-confined realm and creates the new note on the background thread.
- A completion handler is called back to the view controller and the main thread is switched to in the controller for calling UI update functions.
- The table of notes now needs to be updated, so the storage service is called again using the main thread. 
- But before we fetch the data from realm, we need to sync the newer data in the background thread, with our main thread, otherwise we will get stale data back (newest note won't be there).
- Now we can retrieve all notes from the realm on the main thread in our service and pass the data to the view controller to display in the UI.

I think realm is more designed to be called within view controllers/UI logic since its seems to be very performance orientated in general. I'm not sure if there are many other options. Let me know what you think.

wei-lee commented:
@TommyJ1994  will [ThreadSafeReference](https://realm.io/docs/swift/latest#passing-instances-across-threads) help to pass the lists between the background and main threads?

tomjackman commented:
@wei-lee Yes, I was playing around with these before. These could be used (at least for passing a single realm result object, not sure on a list yet - I'll need to test that), however we would need to perform the switch back to the main thread within the storage service to return data to the view controller. Unless we want to define new realm instances in the view controllers and hand down the realm config (encryption key) from the storage service, so we can perform the realm resolve to retrieve the object we marked as thread safe. If its ok to switch threads from the service, I can do it this way. Wasn't quite sure if this was a good practice or not.

wei-lee commented:
@TommyJ1994 Yeah, we probably don't want to define new realm instances in the view controllers, so I think it is ok to switch threads from the service.

In general, it is good practices to invoke the callback in the same thread as the caller. This way it will avoid confusion and the caller can decide if they want to switch to another thread inside the implementation of the callback.

In our case, the VC will invoke the service from the main thread, and it is expected that the callback is running inside the main thread as well. So this means the service itself should handle the thread switch automatically.

tomjackman commented:
@wei-lee Added thread safe calls for DB read note/list notes. After the background thread Read calls are performed on realm, the callback is then called on the main thread from the service.

Also, are we planning to keep the `Pods` directory checked into git? It looks like it will be a bit of a pain with rebases, particulary when new pods are added - some pros/cons of including the Pods directory [here](https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control). It looks like locking down versions/artifacts is the only real benefit of keeping these checked in.

tomjackman commented:
@wei-lee Rebased on master. Retested both Auth and Storage functionality and works as expected.

Its actually not too bad rebasing with the `Pods` folder. I just squashed commits, deleted `Pods/` and `Podfile`, fixed conflicts in `Podfile` and `project.pbxproj`, ran `pod install` and re-opened the project in xcode. I think its ok to keep the `Pods` checked in from my view.
